### PR TITLE
[Paddle] Add chapter chapter_linear-networks

### PR DESCRIPTION
**Pull request by tngt**

### CODE REVIEW

**Comment by tngt** [comment]
hi @AnirudhDagar  submitted the PR according to the provided process, the first PR is used to test process. The second is for committing chap3 content, and the third removes some "None" characters that I didn't notice before.

Now the CI have failed and I find that the problem is that simple_alias is missing in the [library-paddle] section of the cofig.ini file. Can I modify this file?

-----------------
simple_alias = xx
fluent_alias = xx
alias = xx
reverse_alias = xx

**Comment by AnirudhDagar** [comment]
Thanks, @tngt for all the hard work! This seems perfectly organized, as is also reflected in the reduced diff. Yes please go ahead with adding all the aliases to `config.ini`.

File: chapter_linear-networks/image-classification-dataset.md
Comment (AnirudhDagar) [inline chapter_linear-networks/image-classification-dataset.md:1]:
```suggestion
```

File: chapter_linear-networks/linear-regression-concise.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-concise.md:1]:
```suggestion
```

File: chapter_linear-networks/linear-regression-concise.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-concise.md:157]: This block of paddle text should come after the paragraph below which is about tensorflow and keras and should be one whole text tab.

File: chapter_linear-networks/linear-regression-concise.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-concise.md:276]:
```suggestion
```

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-scratch.md:1]:
```suggestion
```

File: chapter_linear-networks/linear-regression.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression.md:1]:
```suggestion
```

File: chapter_linear-networks/softmax-regression-concise.md
Comment (AnirudhDagar) [inline chapter_linear-networks/softmax-regression-concise.md:1]:
```suggestion
```

File: chapter_linear-networks/softmax-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/softmax-regression-scratch.md:1]:
```suggestion
```

File: chapter_linear-networks/softmax-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/softmax-regression-scratch.md:284]: Why do we have this multi-line comment here?

File: chapter_linear-networks/softmax-regression.md
Comment (AnirudhDagar) [inline chapter_linear-networks/softmax-regression.md:1]:
```suggestion
```

File: chapter_linear-networks/softmax-regression.md
Comment (AnirudhDagar) [inline chapter_linear-networks/softmax-regression.md:327]:
```suggestion

```

**Review by AnirudhDagar — CHANGES_REQUESTED** [review]
@tngt I've made a few more suggestions on the PR. Thanks again for working on this.

**Comment by tngt** [comment]
> @tngt I've made a few more suggestions on the PR. Thanks again for working on this.

Thank you for your suggestion. When I load jupyter notebook, it will automatically generate a lot of 'none' characters. I will solve it as soon as possible and PR again. Thank you for your patience and guidance.

**Comment by AnirudhDagar** [comment]
Hi @tngt that's a good question. As mentioned in the [contributing guidelines](https://github.com/d2l-ai/d2l-en/blob/master/CONTRIBUTING.md):

> If the saved functions require some packages to be imported, you can add them to chapter_preface/index.md under the respective framework tab and run d2lbook build lib. Now the import will also be reflected in the d2l library after running and the saved functions can access the imported lib.

In simple words you need to add the imports that you want to save in `paddle.py` file to `chapter_preface/index.md` so that they are included after automatic generation.

You can see my PR here for reference https://github.com/d2l-ai/d2l-zh/pull/1127/files. If you look closely, you'll see the changes made in `chapter_preface/index.md`.

**Comment by tngt** [comment]
> Hi @tngt that's a good question. As mentioned in the [contributing guidelines](https://github.com/d2l-ai/d2l-en/blob/master/CONTRIBUTING.md):
> 
> > If the saved functions require some packages to be imported, you can add them to chapter_preface/index.md under the respective framework tab and run d2lbook build lib. Now the import will also be reflected in the d2l library after running and the saved functions can access the imported lib.
> 
> In simple words you need to add the imports that you want to save in `paddle.py` file to `chapter_preface/index.md` so that they are included after automatic generation.
> 
> You can see my PR here for reference https://github.com/d2l-ai/d2l-zh/pull/1127/files. If you look closely, you'll see the changes made in `chapter_preface/index.md`.

hi, @AnirudhDagar Thank you for your patience, Could you help us review PR and evaluate whether it can be merged？

**Comment by tngt** [comment]
> > Hi @tngt that's a good question. As mentioned in the [contributing guidelines](https://github.com/d2l-ai/d2l-en/blob/master/CONTRIBUTING.md):
> > > If the saved functions require some packages to be imported, you can add them to chapter_preface/index.md under the respective framework tab and run d2lbook build lib. Now the import will also be reflected in the d2l library after running and the saved functions can access the imported lib.
> > 
> > 
> > In simple words you need to add the imports that you want to save in `paddle.py` file to `chapter_preface/index.md` so that they are included after automatic generation.
> > You can see my PR here for reference https://github.com/d2l-ai/d2l-zh/pull/1127/files. If you look closely, you'll see the changes made in `chapter_preface/index.md`.
> 
> hi, @AnirudhDagar Thank you for your patience, Could you help us review PR and evaluate whether it can be merged？

@AnirudhDagar Could you help us review PR and evaluate whether it can be merged？sorry to bother U。

**Comment by AnirudhDagar** [comment]
@tngt I couldn't find the modified `chapter_preface/index.md` file for imports in paddle as I mentioned in my last comment. Could you please commit that as well? I'll then take a pass and request @cheungdaven to also make a final pass before merging. Thanks!

**Comment by tngt** [comment]
> @tngt I couldn't find the modified `chapter_preface/index.md` file for imports in paddle as I mentioned in my last comment. Could you please commit that as well? I'll then take a pass and request @cheungdaven to also make a final pass before merging. Thanks!

It may be that I am doing something wrong, this commit is not displayed but it already exists in the file. Now that I have resubmitted, can you review it again?

> @tngt I couldn't find the modified `chapter_preface/index.md` file for imports in paddle as I mentioned in my last comment. Could you please commit that as well? I'll then take a pass and request @cheungdaven to also make a final pass before merging. Thanks!

@AnirudhDagar Can you help me review it again? thanks

File: chapter_linear-networks/image-classification-dataset.md
Comment (AnirudhDagar) [inline chapter_linear-networks/image-classification-dataset.md:77]: Space after comma in args.

File: chapter_linear-networks/image-classification-dataset.md
Comment (AnirudhDagar) [inline chapter_linear-networks/image-classification-dataset.md:254]: This might break for Mac or Windows.

File: chapter_linear-networks/image-classification-dataset.md
Comment (AnirudhDagar) [inline chapter_linear-networks/image-classification-dataset.md:340]: same as above; line breaks char>80

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-scratch.md:369]: space before and after '='.

**Review by AnirudhDagar — COMMENTED** [review]
@tngt I've made some comments (mostly nits and design specifics and some questions). PTAL. I don't want to block you with this hence will commit the changes required to fastrack the PR.

**Comment by AnirudhDagar** [comment]
Please go through the comments; even though I've resolved most of these. Thanks!

File: chapter_linear-networks/linear-regression-concise.md
Comment (tngt) [reply chapter_linear-networks/linear-regression-concise.md:249]: https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/ParamAttr_cn.html#paramattr

-----
You can click this link to know this API, weight_attr  is initialized in Xavier and bias_attr is initialized in zero bias.

File: chapter_linear-networks/linear-regression-scratch.md
Comment (tngt) [reply chapter_linear-networks/linear-regression-scratch.md:199]: Your suggestion is very useful, we are missing the introduction and I will add the comment that

----

# w和b为读者创建的模型参数，stop_gradient默认为True，即梯度不更新，需要指定为False来更新梯度。

**Comment by w5688414** [comment]
can this PR merge now? it takes us too long and the remaining chapters are ready, we want to finish this pr as soon as possible

File: chapter_linear-networks/linear-regression-concise.md
Comment (cheungdaven) [inline chapter_linear-networks/linear-regression-concise.md:224]: Should this be removed?

File: chapter_linear-networks/linear-regression-concise.md
Comment (cheungdaven) [inline chapter_linear-networks/linear-regression-concise.md:495]: Ditto

**Comment by AnirudhDagar** [comment]
Hi @w5688414,
I'm on the same page with you about getting this in ASAP. But since I can't read Chinese, it is good to have a second review for the comments and also in general maybe some more improvements.

Maybe Shuai was busy this week, I'll request him to review this by today so that we can unblock other chapters.

Appreciate your patience.

File: chapter_linear-networks/linear-regression-scratch.md
Comment (cheungdaven) [inline chapter_linear-networks/linear-regression-scratch.md:367]: condense this comment.

**Comment by cheungdaven** [comment]
@AnirudhDagar  Thanks for the reminder. Hi, @w5688414, the submission seems to be incomplete. Please check my comments.

File: chapter_linear-networks/image-classification-dataset.md
Comment (cheungdaven) [inline chapter_linear-networks/image-classification-dataset.md:258]: I am not sure what happened but it seems the data loading process is much slower than other libraries. It is acceptable though if it cannot be optimized.

File: chapter_linear-networks/softmax-regression-scratch.md
Comment (cheungdaven) [inline chapter_linear-networks/softmax-regression-scratch.md:581]: Is it really a good solution to use global variables here?

**Review by cheungdaven — CHANGES_REQUESTED** [review]

File: chapter_linear-networks/linear-regression-concise.md
Comment (tngt) [reply chapter_linear-networks/linear-regression-concise.md]: Theory has been added

File: chapter_linear-networks/linear-regression-concise.md
Comment (tngt) [reply chapter_linear-networks/linear-regression-concise.md]: removed

**Comment by cheungdaven** [comment]
@w5688414 Could you take a look at the CI issues?

File: chapter_linear-networks/image-classification-dataset.md
Comment (Nietism) [reply chapter_linear-networks/image-classification-dataset.md:258]: Hi! 
As the result shown in http://preview.d2l.ai/d2l-zh/PR-1134/, I noticed that a line in the log is 
```
/home/d2l-worker/miniconda3/envs/d2l-zh-paddle-1/lib/python3.8/site-packages/paddle/fluid/framework.py:311: UserWarning: You are using GPU version Paddle, but your CUDA device is not set properly. CPU device will be used by default.
```
Maybe that is why you observed the data loading process of Paddle is much slower.
Before running the code, invoking ``paddle.device.get_device()`` and ``paddle.device.cuda.device_count()`` might be helpful to locate the problem.

File: d2l/paddle.py
Comment (Nietism) [inline d2l/paddle.py:155]: Should be "Paddle数据迭代器".

File: chapter_linear-networks/image-classification-dataset.md
Comment (cheungdaven) [reply chapter_linear-networks/image-classification-dataset.md:258]: Thanks for the reply. We are using cpus if there are no heavy computation tasks. It is ok to leave it unchanged.

**Comment by w5688414** [comment]
> 

Okay, I am building an offline python environment for testing these md files, so I need more time to deal with python environment problem for supporting MXNET, pytorch, and tf at the same time , but it works fine on our Aistudio(https://aistudio.baidu.com/aistudio/index) platform,  maybe this is the reason for CI test failing

File: d2l/paddle.py
Comment (w5688414) [reply d2l/paddle.py]: Done

**Comment by w5688414** [comment]
@cheungdaven @AnirudhDagar  The SGD bug fixed, can this PR merge now?

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-scratch.md:286]: Is this the right way? What if the params are on a GPU device. In that case will `.numpy()` work? Even if it works, this means we'll be copying data to the CPU from GPU which is not right.

**Comment by cheungdaven** [comment]
@w5688414 Thank you very much. The commits look good to me.

**Review by cheungdaven — APPROVED** [review]

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-scratch.md:285]: Please follow the style guide.
`param -= lr * params[i].grad / batch_size`

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-scratch.md:281]: Extra space after colon.

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [inline chapter_linear-networks/linear-regression-scratch.md:284]: space after comma.
`for i, param in enumerate(params):`

**Review by AnirudhDagar — COMMENTED** [review]
I made a few comments @cheungdaven do you agree with the param.numpy() part specifically?

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [reply chapter_linear-networks/linear-regression-scratch.md:286]: We should probably use something like `param.data` here

**Comment by cheungdaven** [comment]
I do not want to block it for too long but @w5688414  could you please fix the style issues @AnirudhDagar mentioned in later commits. Feel free to create another pr since I have already merged this pr.

File: chapter_linear-networks/linear-regression-scratch.md
Comment (w5688414) [reply chapter_linear-networks/linear-regression-scratch.md:286]: There is no param.data API for PaddlePaddle, but we can change Paddle Tensor's value by using set_value API， I find out the set_value API can receive the Tensor, so the sgd can be:

```
def sgd(params, lr, batch_size):  #@save
    """小批量随机梯度下降"""
    with paddle.no_grad():
        for i, param in enumerate(params):
            param -= lr * params[i].grad / batch_size
            params[i].set_value(param)
            params[i].clear_gradient()
```
is this ok?

File: chapter_linear-networks/linear-regression-scratch.md
Comment (w5688414) [reply chapter_linear-networks/linear-regression-scratch.md:285]: Done

File: chapter_linear-networks/linear-regression-scratch.md
Comment (w5688414) [reply chapter_linear-networks/linear-regression-scratch.md:284]: Done

File: chapter_linear-networks/linear-regression-scratch.md
Comment (AnirudhDagar) [reply chapter_linear-networks/linear-regression-scratch.md:286]: Yes, if this works should be much better.

**Pull request merged**
